ТЗ 12 Групповой проект -> ветка: add-reviews

AndMolPraktikum opened the pull request:
Добавлена модель Review
Добавлен контроллер ReviewController
Добавлен сервисный класс ReviewService
Добавлен интерфейс ReviewStorage
Добавлен DAO класс ReviewDbStorage
Добавлен класс исключения ReviewNotFoundException
В классе ErrorHandler добавлены методы.

AndMolPraktikum committed:
ТЗ 12 Групповой проект -> ветка: add-reviews
Добавлена модель Review
Добавлен контроллер ReviewController
Добавлен сервисный класс ReviewService
Добавлен интерфейс ReviewStorage
Добавлен DAO класс ReviewDbStorage
Добавлен класс исключения ReviewNotFoundException
В классе ErrorHandler добавлены методы.

AndMolPraktikum committed:
ТЗ 12 Групповой проект -> ветка: add-reviews
Добавлены аннотации в тестовые классы FilmControllerTest и UserControllerTest

AndMolPraktikum committed:
ТЗ 12 Групповой проект -> ветка: add-reviews
Удалены неиспользуемые импорты

Sherstyankin commented on src/main/java/ru/yandex/practicum/filmorate/controller/ReviewController.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @PutMapping("/{id}/like/{userId}")
+    public void likeReview(@PathVariable int id, @PathVariable int userId) {
```
Тут id c типом int, а в эндпоинтах ниже с long.

AndMolPraktikum committed:
ТЗ 12 Групповой проект -> ветка: add-reviews
ErrorHandler -> добавлена обработка валидационных исключений
Review -> добавлены аннотации
ReviewController -> добавлен метод кастомный валидатор
ReviewDbStorage -> запросы к БД приведены к общему стандарту
ReviewService -> добавлена проверка userId и filmId
sql_test_data -> добавлены данные тестов.

AndMolPraktikum committed:
ТЗ 12 Групповой проект -> ветка: add-reviews
ReviewControllerTest -> изменены модификаторы доступа

greblovp commented on src/main/java/ru/yandex/practicum/filmorate/controller/ErrorHandler.java:
```diff
@@ -34,6 +34,18 @@ public ErrorResponse handleMPANotFound(final MPANotFoundException e) {
+    @ExceptionHandler
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    public ErrorResponse handleMinReviewValidationException(final MinReviewValidationException e) {
+        return new ErrorResponse("Ошибка в заполнении полей отзыва", e.getMessage());
```
Тут ругаемся на ошибку в заполнении полей, при этом статус NOT_FOUND? Это постман такие условия диктует?)

AndMolPraktikum replied:
![image](https://user-images.githubusercontent.com/116709414/235123604-75bd0071-0f44-4a14-a13b-f282590c8224.png)

greblovp replied:
Кажется, что можно было не делать отдельную проверку на -1. То же самое сделала бы проверка на корректный ID пользователя в методе createReview твоего сервиса...

AndMolPraktikum replied:
Согласен, сейчас прогоню этот момент еще раз.

AndMolPraktikum replied:
По сути проверку на Null тоже можно было убрать. Постман просит 400-й или 500-й статусы. Но так красивее. Проверку на отрицательное значение удалил.

greblovp commented on src/main/java/ru/yandex/practicum/filmorate/dao/impl/ReviewDbStorage.java:
```diff
@@ -0,0 +1,140 @@
+                "\t\tr.is_positive AS isPositive,\n" +
+                "\t\tr.creator_user_id AS userId,\n" +
+                "\t\tr.reviewed_film_id AS filmId,\n" +
+                "\t\tCOALESCE(SUM(rl.score), 0) AS useful\n" +
```
А вот эти спец символы отступов и переносов строк обязательны?

AndMolPraktikum replied:
так понятнее, на мой взгляд. Они не мешают.

greblovp commented on src/main/java/ru/yandex/practicum/filmorate/dao/impl/ReviewDbStorage.java:
```diff
@@ -0,0 +1,140 @@
+                "FROM review AS r LEFT JOIN review_like AS rl ON r.review_id = rl.review_id\n" +
+                "GROUP BY r.review_id, r.content, r.is_positive, r.creator_user_id, r.reviewed_film_id\n" +
+                "ORDER BY 6 DESC, 1\n" +
+                "LIMIT ?;";
```
В getReviewsByFilmId сортировка у тебя указана как "ORDER BY useful DESC, reviewId\n" 
Здесь же используешь указатели на номер столбца. Может выбрать один из вариантов и использовать везде его?

AndMolPraktikum replied:
Хотелось попробовать поразному

greblovp commented on src/main/java/ru/yandex/practicum/filmorate/model/Review.java:
```diff
@@ -0,0 +1,30 @@
+    private String content;  //содержание отзыва
+    @NotNull(message = "review notnull")
+    @JsonProperty("isPositive")
+    private Boolean isPositive;  //тип отзыва
```
Без @JsonProperty("isPositive") не работает?

AndMolPraktikum replied:
Нет, на выходе поле "positive". Работает только с аннотацией.

greblovp reviewed (approved):
Это непростая фича, но все сделано супер.
Понравилось, что удалось унифицировать работу с лайками и дизлайками отзывов.

Я оставил небольшие комментарии, они больше для понимания и возможные правки по стилистике.

Sherstyankin commented on src/main/java/ru/yandex/practicum/filmorate/dao/impl/ReviewDbStorage.java:
```diff
@@ -0,0 +1,140 @@
+        final List<Review> reviews = jdbcTemplate.query(sqlQuery, (rs, rowNum) -> makeReview(rs), reviewId);
+        if (reviews.size() != 1) {
+            log.error("Отзыв с ID {} не существует", reviewId);
+            throw new ReviewNotFoundException(String.format("Отзыв с ID %d не существует", reviewId));
```
Я бы пробросил обработку исключения в сервис, но это не принципиально.

AndMolPraktikum replied:
Я решил оставить иенно здесь, потому что это вроде как валидация входящего объекта.

AndMolPraktikum committed:
ТЗ 12 Групповой проект -> ветка: add-reviews
Удалена проверка на минимальное значение в модели Review и связанное с ней исключение MinReviewValidationException

vladsav33 reviewed (approved):
Посмотрел, замечаний нет
